File: unattended/observer.py
"""共享观察逻辑：读 runstate/events.jsonl + snapshot.json，计算运行状态与统计。

被 unattended/cli.py（status/stats/watch）与 loop 周期状态块复用，
保证统计口径一致。

runstate 根目录与 Config.state_dir 同源（默认 %APPDATA%\\curloop\\runstate）。
"""
from __future__ import annotations

import datetime
import json
import time
from pathlib import Path

from .config import USER_CONFIG_DIR, current_branch

_cache = {"mtime": 0.0, "events": [], "path": ""}


def _slug(name: str) -> str:
    import re

    return re.sub(r'[\\/:*?"<>|]', "_", name).replace(" ", "_") or "default"


def _state_key(project: str) -> str:
    """runstate 目录 key：<项目名>@<分支名>（与 config.project_state_dir 一致）。"""
    p = Path(project)
    return _slug(f"{p.name}@{current_branch(p)}")


def runstate_root(state_dir: Path | None = None) -> Path:
    """默认与 Config.state_dir 一致：%APPDATA%\\curloop\\runstate。"""
    return Path(state_dir) if state_dir is not None else (USER_CONFIG_DIR / "runstate")


def events_path(project: str | None = None, state_dir: Path | None = None) -> Path:
    """events.jsonl for a (project, branch) pair; None = most recently active."""
    root = runstate_root(state_dir)
    if project:
        return root / _state_key(project) / "events.jsonl"
    best, best_mt = root / "events.jsonl", 0.0  # 回退旧全局
    try:
        for p in root.glob("*/events.jsonl"):
            try:
                mt = p.stat().st_mtime
                if mt > best_mt:
                    best, best_mt = p, mt
            except Exception:
                pass
    except Exception:
        pass
    return best


def snapshot_path(project: str | None = None, state_dir: Path | None = None) -> Path:
    return events_path(project, state_dir=state_dir).with_name("snapshot.json")


def load_events(project: str | None = None, state_dir: Path | None = None) -> list[dict]:
    """Read events.jsonl (per project), cached by mtime."""
    p = events_path(project, state_dir=state_dir)
    try:
        mt = p.stat().st_mtime
        if mt != _cache["mtime"] or str(p) != _cache["path"]:
            evs: list[dict] = []
            with p.open(encoding="utf-8") as fh:
                for line in fh:
                    try:
                        evs.append(json.loads(line))
                    except Exception:
                        pass
            _cache.update(mtime=mt, events=evs, path=str(p))
        return _cache["events"]
    except Exception:
        return _cache["events"]


def load_snapshot(project: str | None = None, state_dir: Path | None = None) -> dict:
    try:
        return json.loads(snapshot_path(project, state_dir=state_dir).read_text(encoding="utf-8"))
    except Exception:
        return {}


def fmt_ts(ts: float) -> str:
    try:
        return datetime.datetime.fromtimestamp(ts).strftime("%H:%M:%S")
    except Exception:
        return "?"


def short_detail(e: dict) -> str:
    for k in ("detail", "reason", "task"):
        v = e.get(k)
        if v:
            return str(v)[:80]
    return ""


def _todo_aligned_queue(project: str | None, snap: dict, state_dir: Path | None = None) -> list[dict]:
    """状态展示的队列与 RunState.load 的 done 过滤对齐。

    快照里的任务若在当前 TODO.md 已勾选（[x]），标记 done——否则状态块显示
    pending/running 而真实内存队列在 load() 里已把它们过滤（误导：看起来有
    任务可跑，实际队列空）。读最近 run_start 事件的 todo 字段定位 TODO.md；
    无 todo 信息或文件缺失时原样返回快照队列。
    """
    raw = snap.get("queue", [])
    if not raw:
        return []
    todo: Path | None = None
    for e in load_events(project, state_dir=state_dir):
        if e.get("event") == "run_start" and e.get("todo"):
            todo = Path(e["todo"])
            break
    if not todo or not todo.exists():
        return [
            {
                "text": t.get("text", "")[:70],
                "status": t.get("status"),
                "retries": t.get("retries", 0),
                "switch_reason": t.get("switch_reason"),
            }
            for t in raw
        ]
    from .todo_queue import _norm, parse_all

    done_norms = {t.normalized() for t in parse_all(todo) if t.done}
    out = []
    for t in raw:
        full = t.get("text", "")
        item = {
            "text": full[:70],
            "status": t.get("status"),
            "retries": t.get("retries", 0),
            "switch_reason": t.get("switch_reason"),
        }
        if item["status"] != "done" and _norm(full) in done_norms:
            item["status"] = "done"
        out.append(item)
    return out


def build_status(
    project: str | None = None,
    state_dir: Path | None = None,
) -> dict:
    """Compute stats + recent events + queue from the runstate files.

    project: target project dir (its events/snapshot); None = most recent.
    state_dir: runstate root (default %APPDATA%\\curloop\\runstate).
    """
    evs = load_events(project, state_dir=state_dir)
    snap = load_snapshot(project, state_dir=state_dir)
    st = {
        "switches": 0, "switch_ok": 0, "switch_failed": 0, "emails": [],
        "sends": 0, "tasks_done": 0, "tasks_start": 0, "extend_ok": 0,
        "run_start": None, "run_end": None, "run_end_kind": None,
        "mode": None, "project": None,
    }
    for e in evs:
        ev = e.get("event")
        if ev == "run_start":
            # 每次 run 独立统计：新一轮 run 开始重置所有计数（换号/对话/完成/
            # 账号），不跨 run 累计——用户要求每次开启对话的统计互不混用。
            st = {
                "switches": 0, "switch_ok": 0, "switch_failed": 0, "emails": [],
                "sends": 0, "tasks_done": 0, "tasks_start": 0, "extend_ok": 0,
                "run_start": None, "run_end": None, "run_end_kind": None,
                "mode": None, "project": None,
            }
            st["run_start"] = e.get("ts", 0)
            st["mode"] = e.get("mode")
            st["project"] = e.get("project")
        elif ev in ("interrupt", "run_done", "run_abort"):
            st["run_end"] = e.get("ts", 0)
            st["run_end_kind"] = ev
        elif ev == "switch_start":
            st["switches"] += 1
        elif ev == "switch_ok":
            st["switch_ok"] += 1
            if e.get("email"):
                st["emails"].append(e["email"])
        elif ev == "switch_failed":
            st["switch_failed"] += 1
        elif ev == "sent":
            st["sends"] += 1
        elif ev == "task_done":
            st["tasks_done"] += 1
        elif ev == "task_start":
            st["tasks_start"] += 1
        elif ev == "extend_result" and e.get("new_tasks", 0) > 0:
            st["extend_ok"] += 1

    running = "-"
    if st["run_start"]:
        if st["run_end"]:
            kinds = {"interrupt": "中断", "run_done": "完成", "run_abort": "中止"}
            running = f"已停止({kinds.get(st['run_end_kind'], st['run_end_kind'])})"
        else:
            running = f"{int(time.time() - st['run_start'])}s"

    recent = [
        {"t": fmt_ts(e.get("ts", 0)), "event": e.get("event"), "detail": short_detail(e)}
        for e in reversed(evs[-30:])
    ]
    queue = _todo_aligned_queue(project, snap, state_dir=state_dir)
    return {
        "stats": st, "recent": recent, "queue": queue,
        "running": running, "now": fmt_ts(time.time()),
    }

File: src/observer.ts
import * as fs from 'fs';
import * as path from 'path';
import { USER_CONFIG_DIR, projectStateKey } from './config';
import { EVENT_LOG_KEEP, eventLogPaths } from './runState';
import { norm, parseAll } from './todoQueue';

/** 共享观察逻辑：读 runstate/events.jsonl + snapshot.json，计算运行状态与统计。
 *
 * 被 cli（status/stats/watch）与 loop 周期状态块复用，保证统计口径一致。
 * runstate 根目录与 Config.stateDir 同源（默认 %APPDATA%\curloop\runstate）。
 */

interface CacheEntry {
  mtimeKey: string;
  events: Record<string, unknown>[];
  path: string;
}
let cache: CacheEntry = { mtimeKey: '', events: [], path: '' };

/** 与 config.projectStateDir / projectStateKey 同源。 */
function stateKey(project: string): string {
  return projectStateKey(path.resolve(project));
}

/** 默认与 Config.stateDir 一致：%APPDATA%\curloop\runstate。 */
export function runstateRoot(stateDir?: string): string {
  return stateDir ?? path.join(USER_CONFIG_DIR, 'runstate');
}

/** events.jsonl for a (project, branch, path) pair; undefined = most recently active. */
export function eventsPath(project?: string | null, stateDir?: string): string {
  const root = runstateRoot(stateDir);
  if (project) return path.join(root, stateKey(project), 'events.jsonl');
  let best = path.join(root, 'events.jsonl'); // 回退旧全局
  let bestMt = 0.0;
  try {
    for (const p of fs.readdirSync(root)) {
      const full = path.join(root, p, 'events.jsonl');
      try {
        const mt = fs.statSync(full).mtimeMs;
        if (mt > bestMt) {
          best = full;
          bestMt = mt;
        }
      } catch {
        /* not a runstate dir */
      }
    }
  } catch {
    /* no runstate dir yet */
  }
  return best;
}

export function snapshotPath(project?: string | null, stateDir?: string): string {
  return path.join(path.dirname(eventsPath(project, stateDir)), 'snapshot.json');
}

function mtimeKey(paths: string[]): string {
  const parts: string[] = [];
  for (const p of paths) {
    try {
      const st = fs.statSync(p);
      parts.push(`${p}:${st.mtimeMs}:${st.size}`);
    } catch {
      parts.push(`${p}:missing`);
    }
  }
  return parts.join('|');
}

/** Read events.jsonl + 轮转段 (.1…N)，按时间从旧到新合并；按各文件 mtime 缓存。 */
export function loadEvents(project?: string | null, stateDir?: string): Record<string, unknown>[] {
  const p = eventsPath(project, stateDir);
  const paths = eventLogPaths(p, EVENT_LOG_KEEP);
  if (!paths.length) {
    // 事件文件已不存在（第一次运行 / 已被清理）：不能返回旧缓存——
    // 否则清理 runstate 或切换项目后，旧账号/旧事件会一直显示。
    cache = { mtimeKey: '', events: [], path: p };
    return [];
  }
  try {
    const key = mtimeKey(paths);
    if (key !== cache.mtimeKey || p !== cache.path) {
      const evs: Record<string, unknown>[] = [];
      for (const fp of paths) {
        try {
          for (const line of fs.readFileSync(fp, 'utf-8').split('\n')) {
            if (!line.trim()) continue;
            try {
              evs.push(JSON.parse(line) as Record<string, unknown>);
            } catch {
              /* skip bad line */
            }
          }
        } catch {
          /* file gone */
        }
      }
      cache = { mtimeKey: key, events: evs, path: p };
    }
    return cache.events;
  } catch {
    return cache.events;
  }
}

export function loadSnapshot(project?: string | null, stateDir?: string): Record<string, unknown> {
  try {
    return JSON.parse(fs.readFileSync(snapshotPath(project, stateDir), 'utf-8')) as Record<string, unknown>;
  } catch {
    return {};
  }
}

export function fmtTs(ts: unknown): string {
  try {
    const n = Number(ts);
    if (!Number.isFinite(n) || n <= 0) return '?';
    const d = new Date(n * 1000);
    const pad = (x: number) => String(x).padStart(2, '0');
    return `${pad(d.getHours())}:${pad(d.getMinutes())}:${pad(d.getSeconds())}`;
  } catch {
    return '?';
  }
}

export function shortDetail(e: Record<string, unknown>): string {
  for (const k of ['detail', 'reason', 'task']) {
    const v = e[k];
    if (v) return String(v).slice(0, 80);
  }
  return '';
}

/** 状态展示的完整 TODO 列表：直接解析 TODO.md（含已完成 + 待办），
 *  并用 snapshot 的 running 状态标记当前正在执行的任务。 */
function todoAlignedQueue(
  project: string | null | undefined,
  snap: Record<string, unknown>,
  stateDir?: string,
): Array<{ text: string; status: string | undefined; retries: number; switch_reason: unknown }> {
  // 找到最近一次 run 使用的 TODO 文件
  let todo: string | null = null;
  for (const e of loadEvents(project, stateDir)) {
    if (e['event'] === 'run_start' && e['todo']) todo = String(e['todo']); // 保留最后一次
  }
  const todoFile = todo && fs.existsSync(todo) ? todo : (project ? path.join(project, 'TODO.md') : null);

  // 完整列表来自 TODO.md（含 [x] 已完成），顺序 = 文件顺序
  const tasks = todoFile && fs.existsSync(todoFile) ? parseAll(todoFile) : [];
  if (!tasks.length) {
    // 无 TODO.md：回退 snapshot 队列（至少展示运行中的状态）
    const raw = (snap['queue'] as Record<string, unknown>[]) || [];
    return raw.map((t) => ({
      text: String(t['text'] ?? '').slice(0, 70),
      status: t['status'] as string | undefined,
      retries: Number(t['retries'] ?? 0),
      switch_reason: t['switch_reason'],
    }));
  }

  // snapshot 中 running 的任务 → 标记为进行中（按规范化文本匹配）
  const runningNorms = new Set(
    ((snap['queue'] as Record<string, unknown>[]) || [])
      .filter((t) => t['status'] === 'running' || t['status'] === 'pending')
      .map((t) => norm(String(t['text'] ?? ''))),
  );

  return tasks.map((t) => {
    const isRunning = runningNorms.has(t.normalized());
    const status = t.done ? 'done' : isRunning ? 'running' : 'pending';
    return { text: t.text, status, retries: t.retries, switch_reason: null };
  });
}

export interface StatusStats {
  switches: number;
  switch_ok: number;
  switch_failed: number;
  emails: string[];
  sends: number;
  tasks_done: number;
  tasks_start: number;
  extend_ok: number;
  run_start: number | null;
  run_end: number | null;
  run_end_kind: string | null;
  mode: string | null;
  project: string | null;
}

export interface StatusReport {
  stats: StatusStats;
  recent: Array<{ t: string; event: unknown; detail: string }>;
  queue: Array<{ text: string; status: string | undefined; retries: number; switch_reason: unknown }>;
  running: string;
  now: string;
}

function freshStats(): StatusStats {
  return {
    switches: 0,
    switch_ok: 0,
    switch_failed: 0,
    emails: [],
    sends: 0,
    tasks_done: 0,
    tasks_start: 0,
    extend_ok: 0,
    run_start: null,
    run_end: null,
    run_end_kind: null,
    mode: null,
    project: null,
  };
}

/** Compute stats + recent events + queue from the runstate files. */
export function buildStatus(project?: string | null, stateDir?: string): StatusReport {
  const evs = loadEvents(project, stateDir);
  const snap = loadSnapshot(project, stateDir);
  let st: StatusStats = freshStats();
  for (const e of evs) {
    const ev = e['event'];
    if (ev === 'run_start') {
      // 每次 run 独立统计：新一轮 run 开始重置所有计数，不跨 run 累计
      st = freshStats();
      st.run_start = Number(e['ts'] ?? 0) || null;
      st.mode = e['mode'] === undefined ? null : String(e['mode']);
      st.project = e['project'] === undefined ? null : String(e['project']);
    } else if (ev === 'interrupt' || ev === 'run_done' || ev === 'run_abort') {
      st.run_end = Number(e['ts'] ?? 0) || null;
      st.run_end_kind = String(ev);
    } else if (ev === 'switch_start') {
      st.switches += 1;
    } else if (ev === 'switch_ok') {
      st.switch_ok += 1;
      if (e['email']) st.emails.push(String(e['email']));
    } else if (ev === 'switch_failed') {
      st.switch_failed += 1;
    } else if (ev === 'sent') {
      st.sends += 1;
    } else if (ev === 'task_done') {
      st.tasks_done += 1;
    } else if (ev === 'task_start') {
      st.tasks_start += 1;
    } else if (ev === 'extend_result' && Number(e['new_tasks'] ?? 0) > 0) {
      st.extend_ok += 1;
    }
  }

  let running = '-';
  if (st.run_start) {
    if (st.run_end) {
      const kinds: Record<string, string> = { interrupt: '中断', run_done: '完成', run_abort: '中止' };
      running = `已停止(${kinds[st.run_end_kind ?? ''] ?? st.run_end_kind ?? ''})`;
    } else {
      running = `${Math.floor(Date.now() / 1000 - st.run_start)}s`;
    }
  }

  const recent = evs
    .slice(-30)
    .reverse()
    .map((e) => ({ t: fmtTs(e['ts']), event: e['event'], detail: shortDetail(e) }));
  const queue = todoAlignedQueue(project, snap, stateDir);
  return {
    stats: st,
    recent,
    queue,
    running,
    now: fmtTs(Date.now() / 1000),
  };
}
